driver-core/src/main/com/mongodb/internal/async/client/AsyncMongoCollectionImpl.java

```diff
@@ -1105,7 +1104,7 @@ private InsertOneResult toInsertOneResult(final com.mongodb.bulk.BulkWriteResult
     private InsertManyResult toInsertManyResult(final com.mongodb.bulk.BulkWriteResult result) {
         if (result.wasAcknowledged()) {
             return InsertManyResult.acknowledged(result.getInserts().stream()
-                    .collect(Collectors.toMap(BulkWriteInsert::getIndex, BulkWriteInsert::getId)));
+                    .collect(HashMap::new, (m, v) -> m.put(v.getIndex(), v.getId()), HashMap::putAll));
```

Comment:
This is the fix to the perf errors - caused by inserting multiple RawBsonDocuments - I'd only tested an insertOne.